iothub_client/src/iothubtransport_mqtt_common.c

```diff
@@ -1750,42 +1780,49 @@ static void processIncomingMessageNotification(PMQTTTRANSPORT_HANDLE_DATA transp
     else if ((type == IOTHUB_TYPE_EVENT_QUEUE) && (addInputNamePropertyToMsg(IoTHubMessage, topic_resp) != 0))
     {
         LogError("failure adding input name to property.");
+        IoTHubMessage_Destroy(IoTHubMessage);
     }
     else if (extractMqttProperties(IoTHubMessage, topic_resp, transportData->auto_url_encode_decode) != 0)
     {
         LogError("failure extracting mqtt properties.");
+        IoTHubMessage_Destroy(IoTHubMessage);
     }
     else
     {
-        MESSAGE_CALLBACK_INFO* messageData = (MESSAGE_CALLBACK_INFO*)malloc(sizeof(MESSAGE_CALLBACK_INFO));
-        if (messageData == NULL)
+        // TODO: try sending disposition if anything fails.
+        MESSAGE_DISPOSITION_CONTEXT* dispositionContext = createMessageDispositionContext(msgHandle);
+
+        if (dispositionContext == NULL)
         {
-            LogError("malloc failed");
+            LogError("failed creating message disposition context");
+            IoTHubMessage_Destroy(IoTHubMessage);
+        }
+        else if (IoTHubMessage_SetDispositionContext(IoTHubMessage, dispositionContext, destroyMessageDispositionContext) != IOTHUB_MESSAGE_OK)
```

Comment:
I recommend changing how the memory/owership model here works.  Basically MQTT (and other layers) do initial malloc() but then do a de facto type "release()" of memory into the IoTHub message handle.  But because Hub layer didn't do initial malloc(), it means that the free() has to be delegated back to the layer here.  I just worry this really produces too tight a coupling between everything, and cleanup issues tend to crop up.

I'd recommend - fairly strongly actually - keeping with the existing IoTHubMessage pattern.  So we leave IoTHubMesasage itself to make a copy of whatever data there is and the caller only ever has to call IoTHub_MessageDestroy() to make it go away.  

So basically API call turns into something like below, and yes I know my naming of vars is awful.

```c
IoTHubMessage_SetDispositionContext(hubMessage, unsigned const char* data, size_t dataLen) {
    hubMessage->disposition = malloc(dataLen); memcpy(hubMessage->disposition, dataLen);
}
```

So this is super easy on MQTT beacues it's just a flat buffer.  If we keep HTTP and/or AMQP I looked at that code and it's not quite as easy as there's pointers.  But I think you could do something like

```c
struct AMQP_CONTEXT {
   int whateverDataWeNeed;
   char extraData[]
};
AMQP_CONTEXT*ctxt = (AMQP_CONTEXT*)malloc(sizeof(AMQP_CONTEXT) + strlen(extraDataToCopy) + 1);
ctxt->whateverDataWeNeed = intValueStuff;
strcpy(ctxt->extraData, extraDataToCopy);

MessageSetDisposition(hnadle, ctxt, ctxtLenWeMallocdBasedOnAbove);
```

You'll have to be careful because IIRC some compliers don't really like the [] business at end of struct, so it may need to be a [1].  

And yeah I 100% get this is more complicated on the initial allocation and also on the dereference.  But the complexity is isolated to alloc/use, and there's not extra interdpendencies between the layers that will be way harder to debug if anything goes wrong.

Reply:
As we talked offline, this async disposition construct is also implemented for the AMQP transport, which has 2 pieces of data for the disposition to be sent (link name and message id).
If we decide to change this design, we will need to change how amqp passes the disposition context (maybe serialize the 2 parameters and set that into the IOTHUB_MESSAGE_HANDLE, which would be easy to free() when IoTHubMessage_Destroy is called).

However... All that this design is doing is letting the transports to pass a function that knows how to free the disposition context they are setting in the IOTHUB_MESSAGE_HANDLE. There is no other action done by these callbacks. Please reconsider keeping this design, which is not as convoluted as it might first look.

Reply:
ok i've reviewed this can work as-is.

Reply:
Thanks John.